### ISSUE TITLE
[FEATURE] Add dependencies for packages

### ISSUE THREAD

**Issue by vessd** [SuperCuber/dotter#48]
**Is your feature request related to a problem? Please describe.**
It's strange to me that packages cannot depend on other packages. For example, I want to declare a package `sway` that will depend on `waybar`, `wofi`, etc. Yes, I saw you put all the files for `i3` in one package, but I am not happy with this solution. I would prefer to split them into different packages.

**Describe the solution you'd like**
Change the meaning of the `includes` variable in `local.toml`. This variable should include files similar to `local.toml`.
I imagine my config would look something like this:
```toml
# .dotter/include/common.toml
packages = ["bash"]

[files]

[variables]
```
```toml
# .dotter/include/archlinux.toml
includes = [".dotter/include/common.toml"]
packages = ["pacman"]

[files]

[variables]
```
```toml
# .dotter/include/sway.toml
packages = ["sway", "waybar", "wofi"]

[files]

[variables]
```
```toml
# .dotter/local.toml
includes = [".dotter/include/archlinux.toml", ".dotter/include/sway.toml"]

[files]

[variables]
```
As a result, I have `bash`, `pacman`, `sway`, `waybar` and `wofi` packages.
There may be some confusion here as I mean the values in the `packages` array are appended to the end, however, clashes in the [files] and [variables] sections must be overwritten. 

**Describe alternatives you've considered**
Add `[pkgname.deps]` section? I don't like it as I will be creating empty packages that only consist of dependencies.

**Comment by RasmusBertell** [comment]
I would prefer to have a dependency section which could be defined in `.dotter/global.toml`.
It wouldn't break backwards compatibility and in my opinion would make syntactically more sense.

An example from my dotfiles:

```toml
[default.variables]
username = "rasmus"
locale   = "fi-FI"
history  = 2147483647

[nvim.files]
"nvim/init.vim" = "$XDG_CONFIG_HOME/nvim/init.vim"

[nvim.variables]
scrolloff = 5
tabwidth  = 4
tabspace  = true
textwidth = 80
folding   = "syntax"

[tmux]
depends = [ "default" ]

[tmux.files]
"tmux/tmux.conf" = "$XDG_CONFIG_HOME/tmux/tmux.conf"

[zsh]
depends = [ "default" ]

[zsh.files]
"zsh/zshrc"      = "$XDG_CONFIG_HOME/zsh/.zshrc"
"zsh/zpath"      = "$XDG_CONFIG_HOME/zsh/.zpath"
"zsh/zshenv"     = "$XDG_CONFIG_HOME/zsh/.zshenv"
"zsh/zlogin"     = "$XDG_CONFIG_HOME/zsh/.zlogin"
"zsh/zalias"     = "$XDG_CONFIG_HOME/zsh/.zalias"
"zsh/etc/zshenv" = { target = "/etc/zsh/zshenv", type = "symbolic", owner = "root"}

[task]
depends = [ "default" ]

[task.files]
"task/config" = "$XDG_CONFIG_HOME/task/config"
```

**Comment by vessd** [comment]
@RasmusBertell Looks fine, but I don't understand how can I  override package dependencies in local.toml?
The reason I suggest changing the meaning of `includes` is because it is confusing. When I saw it for the first time, I thought it included files similar to `local.toml`.
And in this case, I should define an empty `archlinux` package (aka virtual in Debina) to declare a list of packages that I will use on this system, which looks weird in my opinion.

**Comment by SuperCuber** [comment]
@vessd @RasmusBertell The problem is that each of you imagines using dependencies in slightly different ways.

With `zsh.depends = ["default"]`, you're having actual dependencies between packages - if `zsh` is enabled, you want `default` to become enabled as well since it uses variables from there.

With `archlinux.depends = ["my", "installed", "packages"]`, you are using it as a "preset" or indeed as a virtual package.

---

The way I envisioned `includes` is different than both of those - I wanted for example `windows.toml` to contain "patches" to packages so that they fit to windows machines, so that on any windows machine I have I can include it.
But, the packages actually needed on the machine should be able to be varied in that case, so even if for example "bash" and "vim" are patched on `windows.toml`, I want to be able to include it but still only deploy "bash" on the machine that doesn't have vim installed.

That is why `included.toml` cannot have a `packages` field - it's not a template for installations, but an optional modification to `global.toml`.

This complexity is why my current solution was "just describe everything in comments and have the user do some manual work when writing `local.toml`", because I didn't want to make a design decision before I actually encountered the need.

---

I think @RasmusBertell 's proposal is the better choice here, for a couple reasons.

Firstly, it's much simpler both to implement and to understand.
There's no clashing possible, there's no infinite recursion possible (if implemented correctly), and it fits the use case that it tries to solve - lifting the algorithmic task of resolving dependencies from the user to the tool.
It's also correct to put it in `global.toml` because a dependency is part of the definition of what a package is - the files of the depending package need the variables from the depended package no matter what machine it's deployed on.

The reason I like @vessd 's proposal less is because not only it breaks that dependency information out of the `global.toml` file, it also results in a duplication - suddenly a whole file is needed for every package that wants dependencies, and understanding the "dependency tree" of a setup requires opening them all and reading them.

Additionally, an example you give for this is an empty `archlinux` package that depends on all the packages that you will use. But that is not how I intended Dotter to be used.
The information on the specific machine should be in `local.toml` and not `global.toml`'s packages - and if you think that all your arch installations will look almost the same, nothing prevents you from having an `archlinux.toml.template` file that Dotter won't read, that you can check into source control and use as a starting point for `local.toml` on a new machine.

**Comment by RasmusBertell** [comment]
Although it is not an intended use case, I think it would still be possible to create an `archlinux` package and use dependency to load specific packages for archlinux.

```toml
# global.toml
[bash.files]
bashrc = "~/.bashrc"

[pacman.files]
"pacman.conf" = { target = "/etc/pacman.conf", type = "symbolic", owner = "root"}

[archlinux]
depends = [ "bash", "pacman" ]
```

```toml
# local.toml
packages = [ "archlinux" ]
```

**Comment by vessd** [comment]
@RasmusBertell Yes, I know, this is exactly what I mean by virtual package.
I suggested a different solution at first, as I have no feeling that package configurations really need to depend on each other.
Let's take `sway` as an example, I want to define it depending on `waybar` and `wofi`. Together they form my desktop environment. But at the same time, this is not true, since they can work independently of each other (especially when you consider the possibility of using templates). For this reason, I have proposed a variant with presets (like virtual in Debian or groups in Arch).
Plus, the `includes` variable initially misled me as I was expecting a C-like `#include`. That is, it will include files similar to local.toml.
